core/vote/vote_manager.go

```diff
@@ -0,0 +1,217 @@
+package vote
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth/downloader"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type (
+	getHighestJustifiedHeaderFunc func(chain consensus.ChainHeaderReader, header *types.Header) *types.Header
+)
+
+// VoteManager will handle the vote produced by self.
+type VoteManager struct {
+	mux *event.TypeMux
+
+	chain       *core.BlockChain
+	chainconfig *params.ChainConfig
+
+	chainHeadCh  chan core.ChainHeadEvent
+	chainHeadSub event.Subscription
+
+	pool    *VotePool
+	signer  *VoteSigner
+	journal *VoteJournal
+
+	engine consensus.PoSA
+
+	getHighestJustifiedHeader getHighestJustifiedHeaderFunc
```

Comment:
No need for this `getHighestJustifiedHeaderFunc` filed at all,  we can use `engine.getHighestJustifiedHeader`

Reply:
Correct. what about made a small modification: not pass this variable into NewVoteManager but only keep this field for more convenient unit test? Because we want to mock valid and invalid func for justified header

Reply:
We can mock `consensus.PoSA`, dont we?

Reply:
Updated